mmseg/datasets/custom.py

```diff
@@ -230,6 +238,66 @@ def get_gt_seg_maps(self):
 
         return gt_seg_maps
 
+    @classmethod
+    def get_classes_and_palette(cls, classes=None, palette=None):
+        """Get class names of current dataset.
+
+        Args:
+            classes (Sequence[str] | str | None): If classes is None, use
+                default CLASSES defined by builtin dataset. If classes is a
+                string, take it as a file name. The file contains the name of
+                classes where each line contains one class name. If classes is
+                a tuple or list, override the CLASSES defined by the dataset.
+        """
+        if classes is None:
+            cls.custom_classes = False
+            return cls.CLASSES, cls.PALETTE
+
+        cls.custom_classes = True
+        if isinstance(classes, str):
+            # take it as a file path
+            class_names = mmcv.list_from_file(classes)
+        elif isinstance(classes, (tuple, list)):
+            class_names = classes
+        else:
+            raise ValueError(f'Unsupported type {type(classes)} of classes.')
+
+        if cls.CLASSES:
+
+            if not classes.issubset(cls.CLASSES):
+                raise ValueError('classes is not a subset of CLASSES.')
+
+            cls.old_id_to_new_id = {}
+            for i, c in enumerate(cls.CLASSES):
+                if c not in class_names:
+                    cls.old_id_to_new_id[i] = -1
+                else:
+                    cls.old_id_to_new_id[i] = classes.index(c)
+
+        return class_names, cls.get_palette_for_custom_classes(palette)
+
+    def get_palette_for_custom_classes(self, palette=None):
+
+        if palette:
+            if isinstance(palette, str):
+                # take it as a file path
+                palette_values = mmcv.list_from_file(palette)
+            elif isinstance(palette, (tuple, list)):
+                palette_values = palette
+            else:
+                raise ValueError(
+                    f'Unsupported type {type(palette)} of palette.')
+
+        else:
+            palette_values = []
+            for x in sorted(self.old_id_to_new_id.items(), key=lambda x: x[1]):
+                if x[1] != -1:
+                    palette_values.append(self.PALETTE[x[0]])
+
+            palette_values = type(self.PALETTE)(palette_values)
+
+        return palette_values
```

Comment:
Custom palette from user input may not be necessary for `CustomDataset`. We may directly set `palette` by indices from subclasses.

Reply:
Yes, sorry. I was still fixing some things... You can now take a look if you can.